src/v/model/tests/timeout_adl.cc

```diff
@@ -35,3 +40,52 @@ BOOST_AUTO_TEST_CASE(test_timeout_clock_is_ms_2) {
     auto res = reflection::adl<model::timeout_clock::duration>{}.from(p);
     BOOST_REQUIRE_EQUAL(res, d);
 };
+
+// In https://github.com/redpanda-data/redpanda/pull/5051
+// the implementation of adl<model::timeout_clock::duration> was refactored
+// This is roughly the old implementation.
+struct adl_legacy_duration {
+    using rep = model::timeout_clock::rep;
+    using duration = model::timeout_clock::duration;
+
+    void to(iobuf& out, duration dur) {
+        // This is a clang bug that cause ss::cpu_to_le to become ambiguous
+        // because rep has type of long long
+        // adl<rep>{}.to(out, dur.count());
+        reflection::adl<uint64_t>{}.to(
+          out,
+          std::chrono::duration_cast<std::chrono::milliseconds>(dur).count());
+    }
+    duration from(iobuf_parser& in) {
+        // This is a clang bug that cause ss::cpu_to_le to become ambiguous
+        // because rep has type of long long
+        // auto rp = adl<rep>{}.from(in);
+        auto rp = reflection::adl<uint64_t>{}.from(in);
+        return std::chrono::duration_cast<duration>(
+          std::chrono::milliseconds{rp});
+    }
+};
+
+constexpr auto durations = std::to_array(
+  {0ms, 16ms, -16ms, 16777216ms, -16777216ms});
+
+// Check that the old implementation and the new implementation are equivalent
+BOOST_DATA_TEST_CASE(test_timeout_clock_legacy, bdata::make(durations), _d) {
+    model::timeout_clock::duration d{_d};
+    // Old serialise, new deserialise
+    {
+        iobuf buf;
+        adl_legacy_duration{}.to(buf, d);
+        iobuf_parser p{buf.share(0, buf.size_bytes())};
+        auto res = reflection::adl<model::timeout_clock::duration>{}.from(p);
+        BOOST_REQUIRE_EQUAL(res, d);
+    }
+    // New serialise, old deserialise
+    {
+        iobuf buf;
+        reflection::adl<model::timeout_clock::duration>{}.to(buf, d);
+        iobuf_parser p{buf.share(0, buf.size_bytes())};
+        auto res = adl_legacy_duration{}.from(p);
+        BOOST_REQUIRE_EQUAL(res, d);
+    }
```

Comment:
what about the case where encode/decode occur on systems with different endian-ness? it seems like there would be a problematic scenario for negatives because it isn't about overflow it's about byte orderings.